### PR TITLE
Use ACLs to allow access to the cf-execd's runagent.socket

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

**Comment by vpodzime** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6131/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6131/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6131/

**Comment by vpodzime** [comment]
RHEL 8 builds (incl. the SELinux policy): [Build Status](https://ci.cfengine.com/job/pr-pipeline/6134/)

**Comment by vpodzime** [comment]
> [Build Status](https://ci.cfengine.com//job/pr-pipeline/6131/)

FR tests retry: [Build Status](https://ci.cfengine.com/job/fr-tests/2274/)
sequential tests retry: [Build Status](https://ci.cfengine.com/job/sequential-tests/1534/)

File: cf-execd/cf-execd.c [hunk 1]
```diff
@@ -599,6 +650,17 @@ static int SetupRunagentSocket()
...
             }
         }
+        if (StringSetSize(execd_config->runagent_allow_users) > 0)
+        {
+            bool success = SetRunagentSocketACLs(sock_info.sun_path,
+                                                 execd_config->runagent_allow_users);
+            if (!success)
+            {
+                Log(LOG_LEVEL_ERR,
+                "Failed to allow runagent_socket_allow_users users access the runagent socket");
```
Comment (larsewi) [inline cf-execd/cf-execd.c:660]: Please align arguments

File: cf-execd/cf-execd.c [hunk 2]
```diff
@@ -553,20 +568,56 @@ static int SetupRunagentSocket()
...
+}
+
+static int SetupRunagentSocket(const ExecdConfig *execd_config)
+{
+    assert(execd_config != NULL);
+
+    int runagent_socket = -1;
+
+    struct sockaddr_un sock_info;
+    int ret = GetRunagentSocketInfo(&sock_info);
```
Comment (larsewi) [inline cf-execd/cf-execd.c:603]: Does it make more sense for the `GetRunagentSocketInfo` function do the `assert(ret > 0)` and truncation testing? Does it not make more sense for this function to return `-1` on error?

It's a bit confusing to me to check for truncation when calling a getter function.

Comment (vpodzime) [reply cf-execd/cf-execd.c:603]: Yup, makas sense. I wasn't really thinking of it as a function, `static inline` is still just a C variant of copy-paste to me :grin:

**Review by larsewi — DISMISSED** [review]
:+1:

**Review by olehermanse — COMMENTED** [review]
Why is the first commit there?

```
Added changelog entry for unexpected consequence of change in and(), …
```

Seems like this should be removed if you rebase to upstream/master.

File: cf-execd/cf-execd.c [hunk 3]
```diff
@@ -946,13 +948,51 @@ static bool ScheduleRun(EvalContext *ctx, Policy **policy, GenericAgentConfig *c
...
+
         *policy = LoadPolicy(ctx, config);
         ExecConfigDestroy(*exec_config);
         ExecdConfigDestroy(*execd_config);
 
         *exec_config = ExecConfigNew(!ONCE, ctx, *policy);
         *execd_config = ExecdConfigNew(ctx, *policy);
 
+#ifndef __MINGW32__
+        if ((RUNAGENT_SOCKET_DIR == NULL) || (!StringEqual_IgnoreCase(RUNAGENT_SOCKET_DIR, "no")))
```
Comment (olehermanse) [inline cf-execd/cf-execd.c:990]: A comment could make this easier to read:

```
// There is a socket dir, either default or specified (not --runagent-socket-dir='no')
```

File: cf-execd/cf-execd.c [hunk 4]
```diff
@@ -946,13 +948,51 @@ static bool ScheduleRun(EvalContext *ctx, Policy **policy, GenericAgentConfig *c
...
+        {
+            if (!StringSetIsEqual(old_runagent_allow_users,
+                                  (*execd_config)->runagent_allow_users))
+            {
```
Comment (olehermanse) [inline cf-execd/cf-execd.c:994]:
```
// List of users in policy has changed
```

(At first I was wondering if old users referred to the old users in ACL on file).

File: cf-execd/cf-execd.c [hunk 5]
```diff
@@ -541,6 +541,16 @@ static void CFExecdMainLoop(EvalContext *ctx, Policy **policy, GenericAgentConfi
             break;
         }
     }
+
+    /* Remove the runagent socket (if any). */
+    if ((RUNAGENT_SOCKET_DIR == NULL) || (!StringEqual_IgnoreCase(RUNAGENT_SOCKET_DIR, "no")))
```
Comment (olehermanse) [inline cf-execd/cf-execd.c:546]: Could be more readable as macro or inline function?

Comment (olehermanse) [reply cf-execd/cf-execd.c:546]:
```C
static inline bool UsesRunAgentSocket()
{
    return ((RUNAGENT_SOCKET_DIR == NULL) || (!StringEqual_IgnoreCase(RUNAGENT_SOCKET_DIR, "no")));
}
```

File: cf-execd/cf-execd.c [hunk 6]
```diff
@@ -602,7 +602,18 @@ static int SetupRunagentSocket(const ExecdConfig *execd_config)
...
+
+        /* Make sure the permissions are correct if the directory was created
+         * (note: this code doesn't run on Windows). */
+        if (created)
+        {
+            char *last_slash = strrchr(sock_info.sun_path, '/');
+            *last_slash = '\0';
+            chmod(sock_info.sun_path, (mode_t) 0700);
+            *last_slash = '/';
+        }
```
Comment (olehermanse) [inline cf-execd/cf-execd.c]: Shouldn't this be done while creating the directory? By setting umask, creating, and then restoring?

**Review by olehermanse — APPROVED** [review]
LGTM - just some smaller comments.

**Comment by vpodzime** [comment]
> Seems like this should be removed if you rebase to upstream/master.

That's exactly what I did and how it got there. :confused:

File: cf-execd/cf-execd.c (hunk 6 above)
Comment (vpodzime) [reply cf-execd/cf-execd.c]: Yes, it should, but that's what `MakeParentDirectory()` is already doing. So I'd have to change it or reimplement it.

Comment (olehermanse) [reply cf-execd/cf-execd.c]: Makes sense 👍

File: cf-execd/cf-execd.c [hunk 7]
```diff
@@ -911,13 +977,52 @@ static bool ScheduleRun(EvalContext *ctx, Policy **policy, GenericAgentConfig *c
...
         *exec_config = ExecConfigNew(!ONCE, ctx, *policy);
         *execd_config = ExecdConfigNew(ctx, *policy);
 
+#ifndef __MINGW32__
+        if (UsingRunagentSocket())
+        {
+            /* Check if the old list and the new one differ. */
+            if (!StringSetIsEqual(old_runagent_allow_users,
+                                  (*execd_config)->runagent_allow_users))
+            {
```
Comment (olehermanse) [inline cf-execd/cf-execd.c:1003]: If I'm being nitpicky: This comment would not help me much. Right now, it says the same as the code. For it to be helpful, I think it needs to mention that the policy changed (not the runagent sockets/dirs/files). For example:
```suggestion
            /* Check if the list of allowed users, from policy, has changed. */
            if (!StringSetIsEqual(old_runagent_allow_users,
                                  (*execd_config)->runagent_allow_users))
            {
```

**Review by olehermanse — APPROVED** [review]

File: cf-execd/cf-execd.c (hunk 7 above)
Comment (vpodzime) [reply cf-execd/cf-execd.c:1003]: I think it's enough to look at where that list is taken from, just a few lines above.

**Pull request merged**